--- /kernel/security/selinux/include/security.h	2016-02-18 20:07:42.000000000 +0000
+++ security.h	2016-04-07 06:36:16.000000000 +0000
@@ -202,8 +202,12 @@
 #define SECURITY_FS_USE_NONE		5 /* no labeling support */
 #define SECURITY_FS_USE_MNTPOINT	6 /* use mountpoint labeling */
 
-int security_fs_use(const char *fstype, unsigned int *behavior,
-	u32 *sid);
+// Vince
+//int security_fs_use(const char *fstype, unsigned int *behavior,
+//	u32 *sid);
+int security_fs_use(const char *fstype, const char *subtype, unsigned int *behavior,
+ 	u32 *sid);
+
 
 int security_genfs_sid(const char *fstype, char *name, u16 sclass,
 	u32 *sid);
